File: mobile/test/operators/test_gru_op.cpp
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include "../test_helper.h"
#include "../test_include.h"
#include "operators/gru_op.h"

namespace paddle_mobile {

template <typename Itype, typename Otype>
int TestGruOp(int in_channels, int out_channels, std::string opname) {
  size_t input_c = in_channels;
  size_t output_c = out_channels;
  paddle_mobile::framework::LoD lod{{0, input_c}};
  int batch_size = lod.size();
  framework::DDim input_shape = framework::make_ddim({input_c, output_c * 3});
  framework::DDim weight_shape = framework::make_ddim({output_c, output_c * 3});
  framework::DDim h0_shape = framework::make_ddim({batch_size, output_c});
  framework::DDim bias_shape = framework::make_ddim({batch_size, output_c * 3});

  VariableNameMap inputs;
  VariableNameMap outputs;
  auto scope = std::make_shared<framework::Scope>();
  inputs["Input"] = std::vector<std::string>({"input"});
  inputs["Weight"] = std::vector<std::string>({"weight"});
  inputs["H0"] = std::vector<std::string>({"h0"});
  inputs["Bias"] = std::vector<std::string>({"bias"});

  outputs["BatchGate"] = std::vector<std::string>({"output_batch_gate"});
  outputs["BatchResetHiddenPrev"] =
      std::vector<std::string>({"output_batch_reset_hidden_prev"});
  outputs["BatchHidden"] = std::vector<std::string>({"output_batch_hidden"});
  outputs["Hidden"] = std::vector<std::string>({"output_hidden"});

  auto input_var = scope.get()->Var("input");
  auto input = input_var->template GetMutable<framework::LoDTensor>();
  SetupTensor<Itype>(input, input_shape, -127, 127);
  input->set_lod(lod);

  auto weight_var = scope.get()->Var("weight");
  auto weight = weight_var->template GetMutable<framework::LoDTensor>();
  SetupTensor<Itype>(weight, weight_shape, -127, 127);

  auto h0_var = scope.get()->Var("h0");
  auto h0 = h0_var->template GetMutable<framework::LoDTensor>();
  SetupTensor<Itype>(h0, h0_shape, -127, 127);

  auto bias_var = scope.get()->Var("bias");
  auto bias = bias_var->template GetMutable<framework::LoDTensor>();
  SetupTensor<Itype>(bias, bias_shape, -127, 127);

  auto batch_gate_var = scope.get()->Var("output_batch_gate");
  auto batch_reset_hidden_prev_var =
      scope.get()->Var("output_batch_reset_hidden_prev");
  auto batch_hidden_var = scope.get()->Var("output_batch_hidden");
  auto hidden_var = scope.get()->Var("output_hidden");

  framework::AttributeMap attrs;
  attrs["activation"].Set<std::string>(std::string("relu"));
  attrs["gate_activation"].Set<std::string>(std::string("sigmoid"));
  attrs["is_reverse"].Set<bool>(false);

  auto *op = new operators::GruOp<CPU, float>("gru", inputs, outputs, attrs,
                                              scope.get());
  op->InferShape();
  op->Init();
  for (int i = 0; i < 10; ++i) {
    op->Run();
  }
  auto time1 = time();
  for (int i = 0; i < 10; ++i) {
    op->Run();
  }
  auto time2 = time();
  std::ofstream out_file("./out_gru.txt", std::ios::app);
  out_file << opname << " cost :" << time_diff(time1, time2) / 10.0 << "ms"
           << std::endl;
  out_file.close();

  delete op;
  return 0;
}

}  // namespace paddle_mobile

int main(int argc, char *argv[]) {
  paddle_mobile::TestGruOp<float, float>(384, 120, "gru_forward");
  return 0;
}

File: lite/tests/unittest_py/op/test_gru_op.py
# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sys
sys.path.append('../')

from auto_scan_test import AutoScanTest, IgnoreReasons
from program_config import TensorConfig, ProgramConfig, OpConfig, CxxConfig, TargetType, PrecisionType, DataLayoutType, Place
import unittest

import hypothesis
from hypothesis import given, settings, seed, example, assume, reproduce_failure
import hypothesis.strategies as st
import argparse
import numpy as np
from functools import partial


class TestGruOp(AutoScanTest):
    def __init__(self, *args, **kwargs):
        AutoScanTest.__init__(self, *args, **kwargs)
        self.enable_testing_on_place(
            TargetType.X86,
            PrecisionType.FP32,
            DataLayoutType.NCHW,
            thread=[1, 4])

        #not support
        # self.enable_testing_on_place(
        #     TargetType.ARM,
        #     PrecisionType.INT8,
        #     DataLayoutType.NCHW,
        #     thread=[1, 4])

        self.enable_testing_on_place(
            TargetType.ARM,
            PrecisionType.FP16,
            DataLayoutType.NCHW,
            thread=[1, 4])

        self.enable_testing_on_place(
            TargetType.ARM,
            PrecisionType.FP32,
            DataLayoutType.NCHW,
            thread=[1, 4])

    def is_program_valid(self,
                         program_config: ProgramConfig,
                         predictor_config: CxxConfig) -> bool:
        return True

    def sample_program_configs(self, draw):
        shape0 = draw(st.integers(min_value=1, max_value=3))
        shape1 = draw(st.integers(min_value=1, max_value=3))
        shape2 = draw(st.integers(min_value=1, max_value=3))
        lod_arr = [0, shape0, shape0 + shape1, shape0 + shape1 + shape2]

        is_rev = draw(st.sampled_from([False, True]))
        bool_orimode = draw(st.sampled_from([True, False]))
        shape_0 = draw(st.integers(min_value=1, max_value=60))
        in_shape = [shape_0, shape_0 * 3]
        batch_num = lod_arr[3]
        h0_1 = len(lod_arr) - 1

        def generate_input(*args, **kwargs):
            return np.random.random(
                [batch_num, in_shape[1]]).astype(np.float32)

        def generate_weight(*args, **kwargs):
            return np.random.random(in_shape).astype(np.float32)

        def generate_bias(*args, **kwargs):
            return np.random.random([1, in_shape[1]]).astype(np.float32)

        def generate_h0(*args, **kwargs):
            return np.random.random([h0_1, in_shape[0]]).astype(np.float32)

        build_ops = OpConfig(
            type="gru",
            inputs={
                "Input": ["input_data"],
                "Weight": ["weight_data"],
                "Bias": ["bias_data"],
                "H0": ["h0"]
            },
            outputs={
                "Hidden": ["hidden"],
                "BatchGate": ["batch_gate"],
                "BatchResetHiddenPrev": ["batch_reset_hidden_prev"],
                "BatchHidden": ["batch_hidden"]
            },
            attrs={
                "activation": "tanh",
                "gate_activation": "sigmoid",
                "is_reverse": is_rev,
                "origin_mode": bool_orimode,
                "is_test": True
            })

        program_config = ProgramConfig(
            ops=[build_ops],
            weights={},
            inputs={
                "input_data": TensorConfig(
                    data_gen=partial(generate_input), lod=[lod_arr]),
                "bias_data": TensorConfig(data_gen=partial(generate_bias)),
                "h0": TensorConfig(data_gen=partial(generate_h0)),
                "weight_data": TensorConfig(data_gen=partial(generate_weight)),
            },
            outputs=["hidden"])
        return program_config

    def sample_predictor_configs(self):
        return self.get_predictor_configs(), ["gru"], (6e-4, 6e-4)

    def add_ignore_pass_case(self):
        def _teller1(program_config, predictor_config):
            if predictor_config.target() == TargetType.ARM:
                if predictor_config.precision() == PrecisionType.INT8:
                    return True

        self.add_ignore_check_case(
            _teller1, IgnoreReasons.PADDLELITE_NOT_SUPPORT,
            "Lite does not support this op for precision int8 on ARM for now.")

    def test(self, *args, **kwargs):
        self.run_and_statis(quant=False, max_examples=100)


if __name__ == "__main__":
    unittest.main(argv=[''])
